APE_Bayesian_modeling.rst

```diff
@@ -0,0 +1,228 @@
+in astronomy, a very common use case is a physical model for the data with an
+explicit data distribution around that model.  This makes it relatively
+straightforward to explicitly express the prior and the likelihood function.
+With an appropriate statistical sampling method (e.g. Markov Chain Monte Carlo),
+these can be used to infer the posterior probability distribution of the
+parameters of the model.  This leads to a more statistically rigorous and
+often more useful version of "fitting a model" . That is, probabilistic
+fitting can result in the best fit model (just like minimizing the existing
+fitting that focuses on minimization of an objective function like chi-squared)
+*and* provide a fully probabilistic description of the model that captures
+uncertainties as fully as the model allows.
+
+However, all of this is not possible with the current implementation of
+``modeling``, as models assume that parameters are a specific value, not a
+distribution. While this makes understanding models simpler, it also makes the
+probabilistic approach to fitting impossible.  This APE aims to remove that
+limitation, while still allowing the existing simpler approach to modeling for
+the existing cases where it is already of clear use.
+
+This APE will also enhance ``fitting`` to allow the use of probabilistic samplers
+that are necessary for Bayesian inference.  Note though, that this APE does
+*not* aim to include samplers in ``astropy`` itself, but rather provide layers that
+interface with samplers already used in the astronomy community like emcee_ or
+dynesty_ . This is akin to the already-existing approach ``fitting`` takes of
+using ``scipy``'s fitters but wrapping them in a more astronomy-friendly
+interface.
+
+Branches and pull requests
+--------------------------
+
+Exploration of some of the ideas in this APE was done in
+https://github.com/astropy/astropy/pull/10839
+
+(PRs implementing portions of this proposal planned)
+
+Implementation
+--------------
+
+The implementation should proceed roughly as ordered below.  Specifically,
+first the :ref:`priors<Priors>`, then the :ref:`statistics<Statistics>` subsections.
+The next two subsections, :ref:`fitters<Fitters>` and :ref:`evidence<Evidence>`
+may be developed in either order.
+
+Priors
+======
+
+Three different ways of specifying a prior on a model should be supported.
+
+* The existing ``bounds`` attribute of ``Parameter`` should be interpreted as
+  a flat prior between the min/max values given in ``bounds``.
+* The ``prior`` attribute of the ``Parameter`` class can be set to any 1D function
+  that takes a single value and returns the prior evaluated at that value.
+  This could be any astropy 1D model or a user written function.
+* For more complicated priors that simultaneously depend on multiple parameters,
+  can be defined as a user written function that takes all the parameters as input
+  and returns the prior evaluated at those parameters.
```

Comment:
I meant, what is the interface to the custom prior function? It still gets called, so what is passed in and out.

Reply:
Yes, these details need to be included.

We had always planned on including these details, the scheduling of an hour to discuss this topic *accelerated* the desire to have this draft APE available for discussion, hence not all the details are present yet.

Reply:
That's what I thought. I just wanted to put in a comment here because it's likely that we go over this document and say "hey, what should we write here" (says Moritz who also thinks that replying to comments is more fun than what he should be doing ;-)  )

Reply:
Regarding the first comment above, if we are dealing with compound models, these parameters get different names in the compound model. Is there a reason those names cannot be used?